tools-local/unix/invoke-extension.sh

```diff
@@ -0,0 +1,117 @@
+    echo "                                          Otherwise, search will attempt to find a version of the tool, which may not be the declared version."
+    echo "  (Optional) tool-path: Path to the tool executable."
+    echo ""
+    echo "Checks if the specified tool has its own implementation of the search or acquire script. If so, invokes the corresponding script. Otherwise, invokes the base implementation."
+    echo ""
+    echo "Example #1"
+    echo "invoke-extension.sh search-tool.sh \"/Users/dotnet/corefx\" cmake """
+    echo "  Searches for CMake, not necessarily the declared version, using the default search scripts located within the repository."
+    echo ""
+    echo "Example #2"
+    echo "invoke-extension.sh acquire-tool.sh \"/Users/dotnet/corefx\" cmake """
+    echo "  Acquires the declared version of CMake, using the default search scripts located within the repository."
+    echo ""
+    echo "Example #3"
+    echo "invoke-extension.sh search-tool.sh \"/Users/dotnet/corefx\" cmake \"/Users/dotnet/MyCustomScripts\" strict"
+    echo "  Searches for the declared version of CMake using the search scripts located in \"/Users/dotnet/MyCustomScripts\"."
+    echo ""
+    echo "Example #4"
+    echo "invoke-extension.sh get-version.sh \"/Users/dotnet/corefx\" cmake "" "" \"/Users/dotnet/corefx/Tools/download/cmake/bin/cmake\" "
+    echo "  Get the version number of CMake executable located at /Users/dotnet/corefx/Tools/download/cmake/bin/cmake\"."
+    echo ""
+}
+
+if [ $# -lt 4 ]; then
+    usage
+    exit 1
+fi
+
+if [ -z "$1" ]; then
+    echo "Argument passed as script-name is empty. Please provide a non-empty string."
+    usage
+    exit 1
+fi
+
+if [ -z "$2" ]; then
+    echo "Argument passed as repository-root is empty. Please provide a non-empty string."
+    usage
+    exit 1
+fi
+
+if [ -z "$3" ]; then
+    echo "Argument passed as tool-name is empty. Please provide a non-empty string."
+    usage
+    exit 1
+fi
+
+if [ ! -z "$4" ] && [ ! -d "$4" ]; then
+    echo "Path specified as override-scripts-folder-path does not exist or is not accessible. Path: $4"
+    usage
+    exit 1
+fi
+
+if [ ! -z "$6" ] && [ ! -f "$6" ]; then
+    echo "Path specified as tool-path does not exist or is not accessible. Path: $6"
+    usage
+    exit 1
+fi
+
+extensionScriptName="$1"
+shift
+
+repoRoot="$1"
+toolName="$2"
+overrideScriptsFolderPath="$3"
+strictToolVersionMatch="$4"
+
+scriptPath="$(cd "$(dirname "$0")"; pwd -P)"
+. "$scriptPath/tool-helper.sh"
+
+
+# Get the appropriate extension script in the folders specified.
+#   1. In OverrideScriptsFolderPath check if the tool overrides base implementation
+#       a. If yes then, return the path of the override
+#       b. If no then, return the path of base implementation
+#   2. If OverrideScriptsFolderPath does not exist then, perform 1.a & 1.b in the default scripts folder.
```

Comment:
In the new helper function `invoke_extension`, the comment is as follows -

>     # Gets the appropriate extension script to invoke.
>     # Searches for an override of the extension script. If an override does not exist then, gets the path to base implementation of the script.

Hopefully this comment clarifies. If it doesn't then please suggest an appropriate comment.